File: shared/color.py
# This Source Code Form is subject to the terms of the GNU General Public
# License, version 3. If a copy of the GPL was not distributed with this file,
# You can obtain one at https://www.gnu.org/licenses/gpl.txt.
import colorsys
import logging
from pprint import pprint
import re

from shared.typeclass import DerivingEq, DerivingAdd, DerivingMul, DerivingSub


log = logging.getLogger('color')


def _clamp(v, low, high):
    return min(max(v, low), high)


def _wrap(v, high):
    return v % high


class Color:
    pass


class BHS(Color, DerivingEq, DerivingAdd, DerivingMul, DerivingSub):
    """Brightness, hue, saturation; as used by Philips Hue."""
    def __init__(self, b_, h_, s_):
        super().__init__()
        self.b_ = b_  # [0,255]
        self.h_ = h_  # [0,64457]
        self.s_ = s_  # [0,255]

    @property
    def b(self):
        return _clamp(int(self.b_), 0, 255)

    @property
    def h(self):
        return _wrap(int(self.h_), 2**16)

    @property
    def s(self):
        return _clamp(int(self.s_), 0, 255)

    def __str__(self):
        return "BHS({0.b}, {0.h}, {0.s})".format(self)

    @classmethod
    def from_rgb(cls, rgb):
        r = rgb.r / 256
        g = rgb.g / 256
        b = rgb.b / 256
        hue, light, sat = colorsys.rgb_to_hls(r, g, b)
        return cls(light * 256, hue * 2**16, sat * 256)


class RGB(Color, DerivingEq, DerivingAdd, DerivingMul, DerivingSub):
    """Red, green, blue triple."""
    def __init__(self, r_, g_, b_):
        super().__init__()
        self.r_ = r_  # [0,255]
        self.g_ = g_  # [0,255]
        self.b_ = b_  # [0,255]

    @property
    def r(self):
        return _clamp(int(self.r_), 0, 255)

    @property
    def g(self):
        return _clamp(int(self.g_), 0, 255)

    @property
    def b(self):
        return _clamp(int(self.b_), 0, 255)

    def __str__(self):
        return "RGB({0.r}, {0.g}, {0.b}) or #{0.r:02X}{0.g:02X}{0.b:02X}".format(self)

    @classmethod
    def from_bhs(cls, bhs: BHS):
        bri = bhs.b / 256
        hue = bhs.h / (2**16)
        sat = bhs.s / 256
        r, g, b = colorsys.hls_to_rgb(hue, bri, sat)
        return cls(r * 256, g * 256, b * 256)


class Mired(Color, DerivingEq, DerivingAdd, DerivingMul, DerivingSub):
    """Mired style color temperature."""
    def __init__(self, ct_):
        super().__init__()
        self.ct_ = ct_

    @property
    def ct(self):
        return _clamp(self.ct_, 153, 500)

    def __str__(self):
        return "Mired({})".format(self.ct)


def parse_css_color(style: str) -> Color:
    style = style.strip().lower()
    if style.startswith('#'):
        if len(style) == 4:
            """Strings of the form #RGB"""
            return RGB(
                int(style[1] + style[1], 16),
                int(style[2] + style[2], 16),
                int(style[3] + style[3], 16)
            )
        elif len(style) == 7:
            """Strings of the form #RRGGBB"""
            return RGB(
                int(style[1:3], 16),
                int(style[3:5], 16),
                int(style[5:7], 16)
            )
        log.error("Mis-formatted color string, expected #RGB or #RRGGBB.")
        raise Exception("Invalid color string, expected #RGB or #RRGGBB.")
    else:
        matches = re.match(r'([a-z]+)\s*\(([0-9,\s]+)\)', style)
        if not matches:
            log.error("Mis-formatted color string.")
            raise Exception("Invalid color string")
        if matches.group(1) == 'rgb':
            """Strings of the form rgb(r,g,b)"""
            parts = matches.group(2).split(',')
            return RGB(int(parts[0].strip()),
                       int(parts[1].strip()),
                       int(parts[2].strip()))
        elif matches.group(1) == 'bhs':
            """Strings of the form bhs(b,h,s)"""
            parts = matches.group(2).strip().split(',')
            return BHS(int(parts[0].strip()),
                       int(parts[1].strip()),
                       int(parts[2].strip()))
        elif matches.group(1) == 'mired':
            """Strings of the form mired(ct)"""
            ct = int(matches.group(2).strip())
            return Mired(ct)
        log.error("Mis-formatted color string, expected type(component...).")
        raise Exception("Invalid color string, expected type(component...)")

File: oh_daemon/src/oh/color.rs
// This Source Code Form is subject to the terms of the GNU General Public
// License, version 3. If a copy of the GPL was not distributed with this file,
// You can obtain one at https://www.gnu.org/licenses/gpl.txt.
use failure::Fallible;
use regex::Regex;

#[derive(Debug, Eq, PartialEq)]
pub struct BHS {
    pub brightness: u8,
    pub hue: u16,
    pub saturation: u8,
}

impl BHS {
    pub fn new(brightness: u8, hue: u16, saturation: u8) -> Fallible<Self> {
        Ok(Self {
            brightness,
            hue,
            saturation,
        })
    }
}

#[derive(Debug, Eq, PartialEq)]
pub struct RGB {
    pub red: u8,
    pub green: u8,
    pub blue: u8,
}

impl RGB {
    pub fn new(red: u8, green: u8, blue: u8) -> Fallible<RGB> {
        Ok(RGB { red, green, blue })
    }
}

#[derive(Debug, Eq, PartialEq)]
pub struct Mired {
    pub color_temp: u8,
}

impl Mired {
    pub fn new(color_temp: u8) -> Fallible<Self> {
        ensure!(
            color_temp >= 40,
            "mired: color temp less than 40 is not meaningful"
        );
        ensure!(
            color_temp <= 200,
            "mired: color temp greater than 200 is not meaningful"
        );
        Ok(Self { color_temp })
    }
}

#[derive(Debug, Eq, PartialEq)]
pub enum Color {
    BHS(BHS),
    RGB(RGB),
    Mired(Mired),
}

impl Color {
    pub fn parse(s: &str) -> Fallible<Color> {
        lazy_static! {
            static ref REGEX_MIRED: Regex = Regex::new(r"\s*mired\(\s*(\d+)\s*\)\s*").unwrap();
            static ref REGEX_RGB: Regex =
                Regex::new(r"\s*rgb\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*\)\s*").unwrap();
            static ref REGEX_BHS: Regex =
                Regex::new(r"\s*bhs\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*\)\s*").unwrap();
        }
        if let Some(captures) = REGEX_BHS.captures(s) {
            return Ok(Color::BHS(BHS::new(
                captures[1].parse()?,
                captures[2].parse()?,
                captures[3].parse()?,
            )?));
        } else if let Some(captures) = REGEX_RGB.captures(s) {
            return Ok(Color::RGB(RGB::new(
                captures[1].parse()?,
                captures[2].parse()?,
                captures[3].parse()?,
            )?));
        } else if let Some(captures) = REGEX_MIRED.captures(s) {
            return Ok(Color::Mired(Mired::new(captures[1].parse()?)?));
        }
        bail!("color: not a color: '{}'", s);
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_parse_mired() -> Fallible<()> {
        assert_eq!(Color::parse("mired(40)")?, Color::Mired(Mired::new(40)?));
        assert_eq!(Color::parse(" mired(40)")?, Color::Mired(Mired::new(40)?));
        assert_eq!(Color::parse("mired(40) ")?, Color::Mired(Mired::new(40)?));
        assert_eq!(Color::parse(" mired(40) ")?, Color::Mired(Mired::new(40)?));
        assert_eq!(Color::parse(" mired( 40) ")?, Color::Mired(Mired::new(40)?));
        assert_eq!(Color::parse(" mired(40 ) ")?, Color::Mired(Mired::new(40)?));
        assert_eq!(
            Color::parse(" mired( 40 ) ")?,
            Color::Mired(Mired::new(40)?)
        );
        assert!(Color::parse("  mired(  39  )  ").is_err());
        assert!(Color::parse("  mired(  201  )  ").is_err());
        assert!(Color::parse("  mired(  256  )  ").is_err());
        Ok(())
    }

    #[test]
    fn test_parse_rgb() -> Fallible<()> {
        assert_eq!(Color::parse("rgb(0,0,0)")?, Color::RGB(RGB::new(0, 0, 0)?));
        assert_eq!(Color::parse(" rgb(0,0,0)")?, Color::RGB(RGB::new(0, 0, 0)?));
        assert_eq!(Color::parse("rgb( 0,0,0)")?, Color::RGB(RGB::new(0, 0, 0)?));
        assert_eq!(Color::parse("rgb(0 ,0,0)")?, Color::RGB(RGB::new(0, 0, 0)?));
        assert_eq!(Color::parse("rgb(0, 0,0)")?, Color::RGB(RGB::new(0, 0, 0)?));
        assert_eq!(Color::parse("rgb(0,0 ,0)")?, Color::RGB(RGB::new(0, 0, 0)?));
        assert_eq!(Color::parse("rgb(0,0, 0)")?, Color::RGB(RGB::new(0, 0, 0)?));
        assert_eq!(Color::parse("rgb(0,0,0 )")?, Color::RGB(RGB::new(0, 0, 0)?));
        assert_eq!(Color::parse("rgb(0,0,0) ")?, Color::RGB(RGB::new(0, 0, 0)?));
        assert_eq!(
            Color::parse(" rgb( 0 , 0 , 0 ) ")?,
            Color::RGB(RGB::new(0, 0, 0)?)
        );
        assert!(Color::parse("rgb(-1,0,0)").is_err());
        assert!(Color::parse("rgb(0,-1,0)").is_err());
        assert!(Color::parse("rgb(0,0,-1)").is_err());
        assert!(Color::parse("rgb(256,0,0)").is_err());
        assert!(Color::parse("rgb(0,256,0)").is_err());
        assert!(Color::parse("rgb(0,0,256)").is_err());
        Ok(())
    }

    #[test]
    fn test_parse_bhs() -> Fallible<()> {
        assert_eq!(Color::parse("bhs(0,0,0)")?, Color::BHS(BHS::new(0, 0, 0)?));
        assert_eq!(Color::parse(" bhs(0,0,0)")?, Color::BHS(BHS::new(0, 0, 0)?));
        assert_eq!(Color::parse("bhs( 0,0,0)")?, Color::BHS(BHS::new(0, 0, 0)?));
        assert_eq!(Color::parse("bhs(0 ,0,0)")?, Color::BHS(BHS::new(0, 0, 0)?));
        assert_eq!(Color::parse("bhs(0, 0,0)")?, Color::BHS(BHS::new(0, 0, 0)?));
        assert_eq!(Color::parse("bhs(0,0 ,0)")?, Color::BHS(BHS::new(0, 0, 0)?));
        assert_eq!(Color::parse("bhs(0,0, 0)")?, Color::BHS(BHS::new(0, 0, 0)?));
        assert_eq!(Color::parse("bhs(0,0,0 )")?, Color::BHS(BHS::new(0, 0, 0)?));
        assert_eq!(Color::parse("bhs(0,0,0) ")?, Color::BHS(BHS::new(0, 0, 0)?));
        assert_eq!(
            Color::parse(" bhs( 0 , 0 , 0 ) ")?,
            Color::BHS(BHS::new(0, 0, 0)?)
        );
        assert!(Color::parse("bhs(-1,0,0)").is_err());
        assert!(Color::parse("bhs(0,-1,0)").is_err());
        assert!(Color::parse("bhs(0,0,-1)").is_err());
        assert!(Color::parse("bhs(256,0,0)").is_err());
        assert!(Color::parse("bhs(0,65566,0)").is_err());
        assert!(Color::parse("bhs(0,0,256)").is_err());
        Ok(())
    }
}
